### PR TITLE
MOI to MatOI

### PR DESCRIPTION
**Pull request by akshay326**
My bad. Took a whole week. But generates matrices!

### CODE REVIEW

**Comment by akshay326** [comment]
Almost the same code as https://github.com/jump-dev/SCS.jl/blob/master/src/MOI_wrapper.jl
tested and works on COSMO, ProxSDP too

File: src/matrix_input.jl [hunk 1]
```diff
@@ -228,3 +229,81 @@ struct MILP{T, LP<:AbstractLPForm{T}}
     lp::LP
     variable_type::Vector{VariableType}
 end
+
+struct ConicForm{T, AT<:AbstractMatrix{T}} <: AbstractConicForm{T}
```
Comment (frapac) [inline src/matrix_input.jl]: that would be nice to add the vector in the templates of `ConicForm`, as we did in #4 . That would allow us to use `ConicForm`  in the GPU :-)

File: src/matrix_input.jl [hunk 2]
```diff
@@ -228,3 +229,81 @@ struct MILP{T, LP<:AbstractLPForm{T}}
...
+    for con in con_idx
+        func = MOI.get(model, MOI.ConstraintFunction(), con)
+        set = MOI.get(model, MOI.ConstraintSet(), con)
+        F = typeof(func)
+        S = typeof(set)
+        __load_constraint(conic, CI{F, S}(CONES_OFFSET[S]), func, set)
+        CONES_OFFSET[S] += cons_offset(set)
+    end
+    
+    # now SCS data shud be allocated
```
Comment (frapac) [inline src/matrix_input.jl]: `shud` -> `should`?

File: src/matrix_input.jl [hunk 3]
```diff
@@ -228,3 +229,81 @@ struct MILP{T, LP<:AbstractLPForm{T}}
...
+    # reorder constraints
+    cis = sort(
+        con_idx, 
+        by = x->CONES[typeof(MOI.get(model, MOI.ConstraintSet(), x))]
+    )
+
+    # extract cones
+    cones = MOI.get(model, MOI.ConstraintSet(), cis)
+
+    return ConicForm{Float64, typeof(A)}(
```
Comment (frapac) [inline src/matrix_input.jl]: Could we specify `Float64` in the template of this function?

File: src/conic_form.jl [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+end
+
+function __allocate_variables(conic::ConicData, nvars::Integer)
+    conic.cone = ConeData()
+    VI.(1:nvars)
+end
+
+function __load_variables(conic::ConicData, nvars::Integer)
+    cone = conic.cone
+    m = cone.f + cone.l + cone.q + cone.s + 3cone.ep + 3cone.ed + 3length(cone.p)
```
Comment (frapac) [inline src/conic_form.jl]: Could we add multiply symbols `*` after the three symbols `3` to improve readibility?
`3cone.ep` -> `3 * cone.ep`

File: src/conic_form.jl [hunk 5]
```diff
@@ -0,0 +1,270 @@
+mutable struct ConeData
+    f::Int # number of linear equality constraints
+    l::Int # length of LP cone
+    q::Int # length of SOC cone
+    qa::Vector{Int} # array of second-order cone constraints
+    s::Int # length of SD cone
+    sa::Vector{Int} # array of semi-definite constraints
+    ep::Int # number of primal exponential cone triples
+    ed::Int # number of dual exponential cone triples
+    p::Vector{Float64} # array of power cone params
```
Comment (frapac) [inline src/conic_form.jl]: Could we move `Float64` in `ConeData` template? Something like
`mutable struct ConeData{T}`
Also, do we want `ConeData` to be mutable?

File: src/conic_form.jl [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    nrows::Dict{Int, Int} # The number of rows of each vector sets, this is used by `constrrows` to recover the number of rows used by a constraint when getting `ConstraintPrimal` or `ConstraintDual`
+    function ConeData()
+        new(0, 0, 0, Int[], 0, Int[], 0, 0, Float64[], Dict{Int, Int}())
+    end
+end
+
+mutable struct ModelData
```
Comment (frapac) [inline src/conic_form.jl]: Same question as with `ConicData` :)

File: src/conic_form.jl [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    n::Int # Number of cols/variables
+    I::Vector{Int} # List of rows
+    J::Vector{Int} # List of cols
+    V::Vector{Float64} # List of coefficients
+    b::Vector{Float64} # constants
+    objective_constant::Float64 # The objective is min c'x + objective_constant
+    c::Vector{Float64}
+end
+
+mutable struct ConicData
```
Comment (frapac) [inline src/conic_form.jl]: At first sight, the exact difference between `ConeData` and `ConicData` looks unclear to me

**Comment by blegat** [comment]
Nice, could you adapt it to be closer to the `matrix.jl` and `geometric.jl` files in https://github.com/jump-dev/SCS.jl/pull/192 ?

**Review by frapac — COMMENTED** [review]
Thanks for this contribution! I have just an overall question: is there some plan to add a function `getLinearForm`, similar in spirit as `getConicForm` but for linear problems? If so, how hard it would be to adapt the function `getConicForm` to the linear case?

File: src/conic_form.jl [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    offset = constroffset(conic, ci)
+    rows = constrrows(s)
+    conic.cone.nrows[offset] = length(rows)
+    i = offset .+ rows
+    b = f.constants
+    # @warn ci
+    # @warn offset
+    # @warn rows
+    if s isa MOI.PositiveSemidefiniteConeTriangle
+        b = scalecoef(rows, b, s)
```
Comment (blegat) [inline src/conic_form.jl]: This is specific to SCS, as you can see in https://github.com/jump-dev/SCS.jl/pull/192/, the plan is to have solvers that need specific preprocessing for some constraints to use the `preprocess` keyword of `copy_to`.

File: src/conic_form.jl [hunk 9]
```diff
@@ -0,0 +1,270 @@
+mutable struct ConeData
+    f::Int # number of linear equality constraints
+    l::Int # length of LP cone
+    q::Int # length of SOC cone
+    qa::Vector{Int} # array of second-order cone constraints
+    s::Int # length of SD cone
```
Comment (blegat) [inline src/conic_form.jl]: This is specific to SCS.
Part of it is in the `num_rows::Vector{Int}` field of the `GeometricConicForm` struct.
Solvers like SCS that need metadata like `qa`, `sa`, ... should do it in `preprocess` like SCS does.

File: src/matrix_input.jl (hunk 2 above)
Comment (akshay326) [reply src/matrix_input.jl]: fixed

File: src/matrix_input.jl (hunk 3 above)
Comment (akshay326) [reply src/matrix_input.jl]: added

File: src/conic_form.jl (hunk 4 above)
Comment (akshay326) [reply src/conic_form.jl]: fixed

File: src/conic_form.jl (hunk 5 above)
Comment (akshay326) [reply src/conic_form.jl]: done

File: src/conic_form.jl (hunk 6 above)
Comment (akshay326) [reply src/conic_form.jl]: yep.  whole data structure parameterized now

File: src/conic_form.jl (hunk 7 above)
Comment (akshay326) [reply src/conic_form.jl]: makes sense. removed merged `ConeData` fields to conicdata

File: src/matrix_input.jl (hunk 1 above)
Comment (akshay326) [reply src/matrix_input.jl]: thats a great idea. adding it now

Comment (akshay326) [reply src/matrix_input.jl]: although i dont how should we edit the matrices A,b,c in CuArray form

**Comment by akshay326** [comment]
> Thanks for this contribution! I have just an overall question: is there some plan to add a function `getLinearForm`, similar in spirit as `getConicForm` but for linear problems? If so, how hard it would be to adapt the function `getConicForm` to the linear case?

we did'nt plan getting the linear case. although it shudn't be much difficult - i tried extracting the matrices given LessThan, EqualTo, GreaterThan constraints

File: src/conic_form.jl [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+        offset = cons_offset(conic)
+        push!(s, _s[i.+(1:offset)])
+        push!(y, _y[i.+(1:offset)])
+        i += offset
+    end
+    return s, y
+end
+
+# Computes conic dimensions
+function constroffset(conic::ConicData{T}, ci::CI{<:MOI.AbstractFunction, MOI.Zeros}) where T
```
Comment (matbesancon) [inline src/conic_form.jl]: A better name would be desirable if possible, `constroffset` gets hard to read.
constraint_offset?

File: src/conic_form.jl [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    return ci.value
+end
+#_allocate_constraint: Allocate indices for the constraint `f`-in-`s`
+# using information in `conic` and then update `conic`
+function _allocate_constraint(conic::ConicData{T}, f, s::MOI.Zeros) where T
```
Comment (matbesancon) [inline src/conic_form.jl]:
```suggestion

"""
    _allocate_constraint

Allocate indices for the constraint `f`-in-`s` using information in `conic` and then update `conic`.
"""
function _allocate_constraint(conic::ConicData{T}, f, s::MOI.Zeros) where T
```

File: src/conic_form.jl (hunk 10 above)
Comment (akshay326) [reply src/conic_form.jl]: yep. editing

File: src/conic_form.jl [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+end
+
+# Vectorized length for matrix dimension n
+sympackedlen(n) = div(n*(n+1), 2)
+# Matrix dimension for vectorized length n
+sympackeddim(n) = div(isqrt(1+8n) - 1, 2)
+trimap(i::Integer, j::Integer) = i < j ? trimap(j, i) : div((i-1)*i, 2) + j
+trimapL(i::Integer, j::Integer, n::Integer) = i < j ? trimapL(j, i, n) : i + div((2n-j) * (j-1), 2)
+function _sympackedto(x, n, mapfrom, mapto)
+    @assert length(x) == sympackedlen(n)
```
Comment (matbesancon) [inline src/conic_form.jl]: Replace the assert with
```julia
length(x) == sympackedlen(n) || throw(DimensionMismatch("error message on dimensions"))
```

File: src/conic_form.jl (hunk 11 above)
Comment (akshay326) [reply src/conic_form.jl]: added

File: src/conic_form.jl (hunk 12 above)
Comment (akshay326) [reply src/conic_form.jl]: adding

File: test/conic_form.jl [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    s = [-5.8304524274385224e-18 1.672022895212547e-17 0.25440785071939953 0.17989351643078233 0.17989351643078233 0.21725123843399946 -0.36765368641915924 0.30723964791089897 0.31108967225522005 -0.3676536864191592 0.2172512384339987]
+    y = [0.5447582105062065 0.3219045549528856 0.45524178943652027 -0.3219045563900619 -0.3219045563900619 1.133337233849821 0.9589717750123417 -0.4552417892860504 1.1333372326447106 0.958971775012335 1.1333372338498173]
+    cones = MOI.AbstractVectorSet[MOI.PositiveSemidefiniteConeTriangle(3), MOI.SecondOrderCone(3), MOI.Zeros(1), MOI.Zeros(1)]
+
+    _s, _y = MatOI.restructure_arrays(s,y,cones)
+
+    @test _s ≈ [[-5.8304524274385224e-18, 1.672022895212547e-17, 0.25440785071939953, 0.17989351643078233, 0.17989351643078233, 0.21725123843399946], [-0.36765368641915924, 0.30723964791089897, 0.31108967225522005], [-0.3676536864191592], [0.2172512384339987]] atol=ATOL rtol=RTOL
+    @test _y ≈ [[0.5447582105062065, 0.3219045549528856, 0.45524178943652027, -0.3219045563900619, -0.3219045563900619, 1.133337233849821], [0.9589717750123417, -0.4552417892860504, 1.1333372326447106], [0.958971775012335], [1.1333372338498173]] atol=ATOL rtol=RTOL
+end
```
Comment (matbesancon) [inline test/conic_form.jl]:
```suggestion
end

```

File: src/matrix_input.jl [hunk 14]
```diff
@@ -228,3 +229,80 @@ struct MILP{T, LP<:AbstractLPForm{T}}
...
+    c::VT
+    A::AT
+    b::VT
+    cones::Vector{<: MOI.AbstractVectorSet}
+end
+
+"""
+Convert a `MathOptInterface` model to `MatrixOptInterface` model
+"""
+function getConicForm(::Type{T}, model::M, con_idx) where {T, M <: MOI.AbstractOptimizer}
```
Comment (matbesancon) [inline src/matrix_input.jl]: Julia tends to use snake case, any specific reason here why not `get_conic_form`?

**Comment by matbesancon** [comment]
is there a specific convention on using double underscore for functions? It doesn't read that well

File: src/matrix_input.jl (hunk 14 above)
Comment (blegat) [reply src/matrix_input.jl]: The JuMP style guide recommends `conic_form` or `get_conic_form`

File: test/conic_form.jl (hunk 13 above)
Comment (akshay326) [reply test/conic_form.jl]: added

File: src/matrix_input.jl (hunk 14 above)
Comment (akshay326) [reply src/matrix_input.jl]: thanks, replaced!

**Comment by akshay326** [comment]
> is there a specific convention on using double underscore for functions? It doesn't read that well

yep removed them now

**Comment by akshay326** [comment]
having a hard time moving from get_conic_form to MOI.copy_to. @blegat can u share some example using MOI.copy_to on GeometricConicForm in action

**Comment by blegat** [comment]
If I understand correctly,  you need something close to `get_conic_form` for DiffOpt. If `copy_to` is given, you can do something like:
```julia
function get_conic_form(::Type{T}, src::MOI.ModelLike, con_idx)
    dest = ConicForm{T, ...}()
    idxmap = MOI.copy_to(dest, src)
    return dest
end
```
except that you only need the constraints in `con_idx`, is that right ? Do you need to filter the constraints for DiffOpt or is a matrix with all constraints sufficient ?

**Comment by akshay326** [comment]
took a lot of time but using `MOI.copy_to` now!
finally I understand that I was using a lot of SCS specific code

File: src/conic_form.jl [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+#         i += offset
+#     end
+#     return s, y
+# end
+
+function get_conic_form(::Type{T}, model::M, con_idx) where {T, M <: MOI.AbstractOptimizer}
+    # reorder constraints
+    cis = sort(
+        con_idx, 
+        by = x->CONES[typeof(MOI.get(model, MOI.ConstraintSet(), x))]
```
Comment (blegat) [inline src/conic_form.jl]: You can use, `_set_type(::MOI.ConstraintSet{F,S}) = S` instead of getting a set just to get its type

File: src/conic_form.jl [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+#     for conic in cones
+#         offset = cons_offset(conic)
+#         push!(s, _s[i.+(1:offset)])
+#         push!(y, _y[i.+(1:offset)])
+#         i += offset
+#     end
+#     return s, y
+# end
+
+function get_conic_form(::Type{T}, model::M, con_idx) where {T, M <: MOI.AbstractOptimizer}
```
Comment (blegat) [inline src/conic_form.jl:37]: What is `con_idx`, can there be different constraints of the same type, in this case, shouldn't you remove duplicates before creating the Tuple?

File: src/conic_form.jl (hunk 15 above)
Comment (akshay326) [reply src/conic_form.jl]: thanks. added!

File: src/conic_form.jl (hunk 16 above)
Comment (akshay326) [reply src/conic_form.jl:37]: yes. removing dupes

File: Project.toml [hunk 17]
```diff
@@ -4,8 +4,11 @@ authors = ["JuMP-dev"]
 version = "0.1.0"
 
 [deps]
+COSMO = "1e616198-aa4e-51ec-90a2-23f7fbd31d8d"
```
Comment (blegat) [inline Project.toml]: Remove these dependency on solvers

File: src/conic_form.jl [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+    y = similar(x)
+    for i in 1:n, j in 1:i
+        y[mapto(i, j)] = x[mapfrom(i, j)]
+    end
+    y
+end
+trimap(i::Integer, j::Integer) = i < j ? trimap(j, i) : div((i-1)*i, 2) + j
+trimapL(i::Integer, j::Integer, n::Integer) = i < j ? trimapL(j, i, n) : i + div((2n-j) * (j-1), 2)
+sympackedLtoU(x, n=sympackeddim(length(x))) = _sympackedto(x, n, (i, j) -> trimapL(i, j, n), trimap)
+sympackedUtoL(x, n) = _sympackedto(x, n, trimap, (i, j) -> trimapL(i, j, n))
```
Comment (blegat) [inline src/conic_form.jl]: Why are these here? This is specific to SCS needing to scale off diagonal entries, do you need this too in DiffOpt?

File: Project.toml (hunk 17 above)
Comment (akshay326) [reply Project.toml]: removed

File: src/conic_form.jl (hunk 18 above)
Comment (akshay326) [reply src/conic_form.jl]: sorry for the late response.
yes, we need at least this part to scale diagonal entries for a PSD cone. my only source of cross checking the actual values was https://github.com/cvxgrp/diffcp, which accepts/returns matrices in SCS format

Comment (blegat) [reply src/conic_form.jl]: It doesn't make much sense to do it by default for the conic form. I would suggest DiffOpt to use something like
https://github.com/jump-dev/SCS.jl/blob/c1e3a8d5469e9f289dd0b3cdd87245792d266317/src/MOI_wrapper.jl#L147
In the future, if the scaled psd cone is brought back then SCS and DiffOpt could use this without any preprocessing function.

Comment (blegat) [reply src/conic_form.jl]: See https://github.com/jump-dev/MathOptInterface.jl/issues/531 for the scaled psd cone

Comment (akshay326) [reply src/conic_form.jl]: ok got it now. will need to correct the tests accordingly!

Comment (blegat) [reply src/conic_form.jl]: > should I check the presence of SCS/Mosek solver and edit the matrices accordingly - so that for every solver matrix A is same

Not sure to understand what you mean

Comment (akshay326) [reply src/conic_form.jl]: forgot to edit the comment 😅 . saw the conic solver info at https://github.com/cvxgrp/cvxpy/tree/master/cvxpy/reductions/solvers/conic_solvers and understood this `trimap` thing is specific to SCS and shouldn't be a part of MatOI

Comment (akshay326) [reply src/conic_form.jl]: @blegat is this PR good to go? or is there anything specific to SCS that I'm missing

**Review by blegat — APPROVED** [review]
Looks good, thanks for the sustained effort on this PR!
